apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Span.java

```diff
@@ -263,43 +263,49 @@ public void beforeEnd(long epochMicros) {
     @Override
     protected void afterEnd() {
         if (transaction != null && transaction.isSpanCompressionEnabled() && parent != null) {
-            Span buffered = parent.bufferedSpan.get();
-            if (parent.isFinished() || !isCompressionEligible()) {
-                if (buffered != null) {
-                    if (parent.bufferedSpan.compareAndSet(buffered, null)) {
-                        this.tracer.endSpan(buffered);
+            Span parentBuffered = parent.bufferedSpan.incrementReferencesAndGet();
+            try {
+                if (parent.isFinished() || !isCompressionEligible()) {
```

Comment:
This code can be reached even if this span is asynchronous to it's parent, correct?
If this is the case, we can't rely on `parent.isFinished()`, because it will always be racy.

So if the code outside this branch relies on the parent NOT being finished, this might cause problems.